get-artifacts/get_artifacts.sh

```diff
@@ -0,0 +1,168 @@
+    usage
+    exit 1
+  fi
+fi
+
+# Retry seems to add a lot of time overhead.  Need to look in to this more.
+# curl_attri=("--retry" "3")
+curl_attri=('-H' 'Accept: application/vnd.github.v3+json')
+if [ -n "$TOKEN" ]; then
+  curl_attri+=('-H' "authorization: Bearer ${TOKEN}")
+fi
+echoerr Curl Parameters: "${curl_attri[@]}"
+
+
+# Conditionally retrieves the artifacts from a single jobs, the input parameter which must correspond to a numeric workflow_run_id in github actions
+# and stores the desired artifacts, contained in a list called "ARTIFACTS", in a subdirectory of TARGET_DIR, equal to the workflow_run_id/artifact_name.
+# Artifacts are always zips.   The zip file is downloaded and optionally unziped in to the job's sub directory in the TARGET_DIR, and then deleted.
+# Should the <job dir>/<artifact_name> already exist, no downloads are attempted, as the assumption would be this function has already populated the
+# artifacts in to the job dir.
+#
+# INPUT:
+#  Parameter 1:  A number workflow_run_id
+#
+# ENVIRONMENT:
+#  TARGET_DIR: A root dir for creation of sub directories for each job
+#  ARTIFACTS: A bash array of artifact names to download from github artifacts workflows for this job.
+#  REPO: Github slug corresponding to a repository
+#  UNZIP: unzip and delete the downloaded artifact?
+#
+# SIDE EFFECTS:
+#  The ${TARGET_DIR}/${Parameter 1} dir id populated with the unzipped contents of the artifacts from the job.
+#
+getArtifactsFromJob() {
+  jobId=$1;
+  echoerr getting artifacts for jobId "$jobId"
+  artifact_info="$(curl "${curl_attri[@]}" "https://api.github.com/repos/${REPO}/actions/runs/${jobId}/artifacts"  2>/dev/null)"
+  for artifact_name in "${ARTIFACTS[@]}"; do
+    if [ ! -d "${TARGET_DIR}/${jobId}/${artifact_name}" ]; then
+      mkdir -p "${TARGET_DIR}/${jobId}/${artifact_name}"
+      download_url_str="$(echo "$artifact_info" | jq '.artifacts[] | select(.name=="'"${artifact_name}"'") .archive_download_url')"
+      if [ -n "$download_url_str" ] && [ "$download_url_str" != "null" ]; then
+        download_url="${download_url_str//\"/}"
+        curl -L "${curl_attri[@]}" "$download_url" -o "${TARGET_DIR}/${jobId}/${artifact_name}/${artifact_name}.zip" 2>/dev/null
+        if [ "$DECOMPRESS" = "true" ]; then
+          unzip -q "${TARGET_DIR}/${jobId}/${artifact_name}/${artifact_name}.zip" -d "${TARGET_DIR}/${jobId}/${artifact_name}/"
+          rm "${TARGET_DIR}/${jobId}/${artifact_name}/${artifact_name}.zip"
+        fi
+      else
+        echoerr Artifact not found on workflow run: "${jobId}" with contents:
+        echoerr "$artifact_info"
+      fi
+    fi
+  done
+  echoerr fetched:
+  echoerr "$(ls -d "${TARGET_DIR}/${jobId}/"*/*)"
+}
+
+#
+# Given a REPO, WORKFLOW and (optional) BRANCH, get HISTORY number of prior run artifacts return output workflow_run_ids, one per line.
+#
+getWorkflow_Run_Ids() {
+  #Always using temp files for curl output to prevent shell mangling.
```

Comment:
Need a space after `#`. (Is there a linter for this?)

Also what does "to prevent shell mangling" mean? Could you provide an example or more information?

Reply:
done, and expanded on comment.